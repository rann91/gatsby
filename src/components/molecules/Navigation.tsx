/* @jsx jsx */
import { graphql, useStaticQuery } from 'gatsby'
import { Fragment, useCallback, useState } from 'react'
import { Close, Flex, jsx, MenuButton } from 'theme-ui'
import { rem } from '../../gatsby-plugin-theme-ui'
import {
  Strapi_ComponentSharedLink,
  Strapi_Navigation
} from '../../typings/graphql'
import Link from '../atoms/Link'

interface Props {
  data: {
    strapi: {
      navigation: Strapi_Navigation
    }
  }
}

export const PureNavigation = ({
  data: {
    strapi: { navigation }
  }
}: Props) => {
  const [open, setOpen] = useState(false)
  const toggleMenu = useCallback(() => setOpen(!open), [open])
  const items = (navigation!.menu || []) as Strapi_ComponentSharedLink[]

  return (
    <Fragment>
      <Flex sx={{ display: [null, null, null, 'none'] }}>
        {open ? (
          <Close aria-label="Close menu" onClick={toggleMenu} />
        ) : (
          <MenuButton aria-label="Open Menu" onClick={toggleMenu} />
        )}
      </Flex>
      <nav
        sx={{
          position: ['fixed', null, null, 'static'],
          top: [rem(65), rem(85), null, 0],
          left: 0,
          width: ['100%'],
          backgroundColor: ['rgba(0, 0, 0, 0.75)', null, null, 'transparent']
        }}>
        <ul
          sx={{
            display: [open ? 'flex' : 'none', null, null, 'flex'],
            flexDirection: ['column', null, null, 'row'],
            paddingTop: [2, null, null, 0],
            paddingBottom: [5, null, null, 0],
            paddingLeft: [5, null, null, 0],
            paddingRight: [5, null, null, 0],
            margin: 0,
            listStyle: 'none',
            backgroundColor: ['white', null, null, 'transparent'],
            '::after': {
              content: '""',
              display: [null, null, null, 'none'],
              position: 'absolute',
              zIndex: -1,
              top: '4px',
              left: 0,
              width: '100%',
              height: 'calc(100% - 4px)',
              boxShadow: [
                open ? '0 0 4px rgba(0, 0, 0, 0.1)' : 'none',
                null,
                null,
                'none'
              ]
            }
          }}>
          {items.map((link: Strapi_ComponentSharedLink) => (
            <li
              key={link.id}
              sx={{
                '& + &': {
                  paddingTop: [2, null, null, 0],
                  paddingLeft: [null, null, null, 5]
                }
              }}>
              <Link
                sx={{
                  fontSize: 0,
                  textTransform: 'uppercase',
                  textDecoration: 'none',
                  color: 'secondary',
                  transition: 'color 0.25s',
                  ':hover': {
                    color: 'text'
                  },
                  '&.active': {
                    fontWeight: 'medium',
                    color: 'text'
                  }
                }}
                link={link}
                onClick={toggleMenu}
              />
            </li>
          ))}
        </ul>
      </nav>
    </Fragment>
  )
}

const Navigation = () => {
  const data = useStaticQuery<{
    strapi: {
      navigation: Strapi_Navigation
    }
  }>(graphql`
    query {
      strapi {
        navigation {
          id
          menu {
            ...Link
          }
        }
      }
    }
  `)

  return <PureNavigation data={data} />
}

export default Navigation
