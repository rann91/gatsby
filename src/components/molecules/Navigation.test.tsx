import { render } from '@testing-library/react'
import React from 'react'
import data from '../../json/navigation.json'
import { PureNavigation as Navigation } from './Navigation'

describe('Navigation', () => {
  it('renders correctly', () => {
    const { container } = render(<Navigation data={data} />)

    expect(container).toMatchSnapshot()
  })
})
