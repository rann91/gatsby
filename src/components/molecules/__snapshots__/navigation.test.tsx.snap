// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navigation renders correctly 1`] = `
<div>
  <div
    class="css-1l0o9ue-PureNavigation"
  >
    <button
      aria-label="Open Menu"
      class="css-wwcqj1"
      title="Menu"
    >
      <svg
        class="css-ipc245"
        fill="currentcolor"
        height="24"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
        />
      </svg>
    </button>
  </div>
  <nav
    class="css-4swsle-PureNavigation"
  >
    <ul
      class="css-13786ft-PureNavigation"
    >
      <li
        class="css-1w5hvub"
      >
        <a
          class="css-1ys58o5"
          data-testid="gatsby-link"
          href="/"
          target="_self"
        >
          Home
        </a>
      </li>
      <li
        class="css-1w5hvub"
      >
        <a
          class="css-1ys58o5"
          data-testid="gatsby-link"
          href="/about"
          target="_self"
        >
          About
        </a>
      </li>
      <li
        class="css-1w5hvub"
      >
        <a
          class="css-1ys58o5"
          data-testid="gatsby-link"
          href="/work"
          target="_self"
        >
          Work
        </a>
      </li>
      <li
        class="css-1w5hvub"
      >
        <a
          class="css-1ys58o5"
          data-testid="gatsby-link"
          href="/team"
          target="_self"
        >
          Team
        </a>
      </li>
      <li
        class="css-1w5hvub"
      >
        <a
          class="css-1ys58o5"
          data-testid="gatsby-link"
          href="/blog"
          target="_self"
        >
          Blog
        </a>
      </li>
      <li
        class="css-1w5hvub"
      >
        <a
          class="css-1ys58o5"
          data-testid="gatsby-link"
          href="/contact"
          target="_self"
        >
          Contact
        </a>
      </li>
    </ul>
  </nav>
</div>
`;
